sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1666,7 +1667,14 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
                   // The update value can access columns from both target and source tables.
                   resolveAssignments(assignments, m, MergeResolvePolicy.BOTH))
               case UpdateStarAction(updateCondition) =>
-                val assignments = targetTable.output.map { attr =>
+                val attrs = if (m.withSchemaEvolution) {
```

Comment:
This is some code needed to handle the UpdateStar and InsertStart cases.  Otherwise, this will be on the new target schema, and would not work if there are some fields not in Source that are in Target table.

Ie, MERGE INTO t USING s.  t=(a int), s=(b int).   New target schema should be (a, b), but Update * should only choose s.b, as there is no column a in s.